File: 백준/Gold/2240. 자두나무/자두나무.cc
#include <iostream>

using namespace std;

int t, w;
int arr[1001];
int dp[1001][31][3];
int ans;

int main() {
    ios_base::sync_with_stdio(0); cin.tie(0);
    cin >> t >> w;
    for (int i = 1; i <= t; i++) {
        cin >> arr[i];
        dp[i][0][1] = dp[i-1][0][1];
        if (arr[i] == 1) dp[i][0][1]++;
    }

    for (int i = 1; i <= t; i++) {
        for (int j = 1; j <= w; j++) {
            for (int k = 1; k <= 2; k++) {
                dp[i][j][k] = max(dp[i-1][j-1][3 - k], dp[i-1][j][k]) + (arr[i] == k ? 1 : 0);
            }
        }
    }

    for (int i = 0; i <= w; i++) {
        ans = max(ans, dp[t][i][1]);
        ans = max(ans, dp[t][i][2]);
    }
    cout << ans;
    return 0;
}

File: 백준/Gold/2240. 자두나무/자두나무.py
# dp

import sys
input = sys.stdin.readline

t, w = map(int, input().split())
arr = [-1]

for i in range(t):
    num = int(input())
    arr.append(num - 1)

dp = [[[0, 0] for j in range(w + 1)] for i in range(t + 1)]

for i in range(1, t + 1):
    dp[i][0][0] = dp[i-1][0][0]
    if arr[i] == 0:
        dp[i][0][0] += 1
    for j in range(1, w + 1):
        for k in range(2):
            dp[i][j][k] = max(dp[i-1][j-1][1-k], dp[i-1][j][k])
            if arr[i] == k:
                dp[i][j][k] += 1

ans = 0
for j in range(w + 1):
    for k in range(2):
        ans = max(ans, dp[t][j][k])

print(ans)